### PR TITLE
Send email notifications when commenting on answers and programming annotations

### LINKED ISSUE #1855: Notification for annotations and answer comments
The notifications for those haven't been implemented yet.

### PR DESCRIPTION
**Pull request by fonglh**
Fixes #1855.

### CODE REVIEW

File: app/views/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.html.slim [hunk 1]
```diff
@@ -0,0 +1,12 @@
+- post = @object
+- answer = post.topic.actable
+- course = topic.course
+- host = course.instance.host
+
+- message.subject = t('.subject', course: course.title, topic: answer.question.display_title)
+
+= simple_format(t('.message',
+                  topic: link_to(answer.question.display_title,
+                                 edit_course_assessment_submission_url(course, answer.question.assessment, answer.submission, host: host)),
```
Comment (weiqingtoh) [inline app/views/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.html.slim]: Should we add in the step param? Maybe can check if its an auto graded assessment first? 

In the future can also add in params for programming code annotations. Can add a todo for that?

File: config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    course:
+      assessment:
+        answer:
+          comment_notifier:
+            replied:
+              user_notifications:
+                email:
+                  subject: '%{course} New comment on %{topic}'
+                  message: >
+                    %{post_author} commented, %{post}
```
Comment (weiqingtoh) [inline config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml]: It just occurred to me that text is HTML so you must need to handle that. Have you tested those?

File: app/views/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.html.slim (hunk 1 above)
Comment (fonglh) [inline app/views/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.html.slim]: Step param is necessary. Will do that and check with the various assessment formats. Yup planning to do code annotations as part of the same issue, but new PR

File: config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml (hunk 2 above)
Comment (fonglh) [inline config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml]: Handle in what way?

Comment (weiqingtoh) [inline config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml]: I think `post.text` is rich text and stored in html format -> `"-1\n\nAgain the function definition is incorrect. You need to compare the real part with real part and the imaginary part with imaginary part."`

I think you need to call `format_html(post.text)`, but probably at the notification partial. I'm not too sure myself, better if you can test posting a rich text post and taking a look at the email.

Comment (fonglh) [inline config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml]: Strangely, our `ApplicationHTMLFormattersHelper` isn't available in the notifier views, so the `format_html` method isn't available. `simple_format` actually calls the one from rails.

File: app/views/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.html.slim (hunk 1 above)
Comment (fonglh) [inline app/views/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.html.slim]: I've added the step param and tried on trainings (with the nav pills) and tabbed missions.

File: config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml (hunk 2 above)
Comment (fonglh) [inline config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml]: `format_html` should not be used as that converts the HTML tags to stuff like `&gt;` for display. We do want this to be interpreted as HTML.

**Comment by fonglh** [comment]
Don't merge yet, going to handle the 1st comment case.

File: config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml (hunk 2 above)
Comment (weiqingtoh) [inline config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml]: Apologies, I think the correct function is `format_text`.

**Comment by fonglh** [comment]
Shall check for notifications on submission level comments.

File: config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml (hunk 2 above)
Comment (fonglh) [inline config/locales/en/notifiers/course/assessment/answer/comment_notifier/replied/user_notifications/email.yml]: Including our HTML helper into the mailer messes up the hyperlink HTML code.

Using the helper and `format_html` makes the message look like it got escaped twice.

Rail's `simple_format` takes care of the `script` tag already.

**Comment by allenwq** [comment]
LGTM 👍

**Comment by fonglh** [comment]
Yes let's merge this in first, so we have fewer emails to send. Then i'll work on sending the backlog of emails.

**Pull request merged**